ui/raidboss/emulator/data/PopupTextAnalysis.ts

```diff
@@ -0,0 +1,289 @@
+
+type ResolverFunc = () => void;
+
+export type EmulatorLogEvent = LogEvent & {
+  detail: {
+    logs: LineEvent[];
+  };
+}
+
+export type EmulatorNetworkLogEvent = EventResponses['LogLine'] & {
+  detail: {
+    logs: LineEvent[];
+  };
+}
+
+interface ResolverStatus {
+  responseType?: string;
+  responseLabel?: string;
+  initialData: DataType;
+  finalData?: DataType;
+  condition?: boolean;
+  response?: undefined;
+  result?: string;
+  delay?: number;
+  suppressed: boolean;
+  executed: boolean;
+  promise?: Promise<void>;
+}
+
+type EmulatorTriggerHelper = TriggerHelper & {
+  resolver?: Resolver;
+};
+
+class Resolver {
+  private promise?: Promise<void>;
+  private run?: ResolverFunc;
+  private delayUntil?: number;
+  private final?: ResolverFunc;
+  private delayPromise?: Promise<void>;
+  private delayResolver?: (value: void | PromiseLike<void>) => void;
+  public triggerHelper?: EmulatorTriggerHelper;
+
+  constructor(public status: ResolverStatus) {}
+
+  async isResolved(log: LineEvent) {
+    if (this.delayUntil) {
+      if (this.delayUntil < log.timestamp) {
+        delete this.delayUntil;
+        if (this.delayResolver)
+          this.delayResolver();
+        await this.delayPromise;
+      } else {
+        return false;
+      }
+    }
+    if (this.promise)
+      await this.promise;
+    if (this.run)
+      this.run();
+    if (this.final)
+      this.final();
+    return true;
+  }
+  setDelay(delayUntil: number): Promise<void> {
+    this.delayUntil = delayUntil;
+    return this.delayPromise = new Promise((res) => {
+      this.delayResolver = res;
+    });
+  }
+  setPromise(promise: Promise<void>) {
+    this.promise = promise;
+  }
+  setRun(run: ResolverFunc) {
+    this.run = run;
+  }
+  setFinal(final: ResolverFunc) {
+    this.final = final;
+  }
+  setHelper(triggerHelper: EmulatorTriggerHelper) {
+    this.triggerHelper = triggerHelper;
+  }
+}
+
+export default class PopupTextAnalysis extends StubbedPopupText {
+  triggerResolvers: Resolver[] = [];
+  currentResolver?: Resolver;
+  public callback?: (log: LineEvent,
+      triggerHelper: EmulatorTriggerHelper | undefined,
+      currentTriggerStatus: ResolverStatus,
+      finalData: DataType) => void;
+
+  // Override `OnTrigger` so we can use our own exception handler
+  OnTrigger(trigger: LooseTrigger, matches: RegExpExecArray | null, currentTime: number): void {
+    try {
+      this.OnTriggerInternal(trigger, matches, currentTime);
+    } catch (e) {
+      console.log(trigger, e);
+    }
+  }
+
+  async OnLog(e: EmulatorLogEvent): Promise<void> {
+    for (const logObj of e.detail.logs) {
+      const log = logObj.properCaseConvertedLine ?? logObj.convertedLine;
+
+      if (log.includes('00:0038:cactbot wipe'))
+        this.SetInCombat(false);
+
+      for (const trigger of this.triggers) {
+        // partial commit, make sure you amend!
+        const r = trigger.localRegex?.exec(log);
+        if (r) {
```

Comment:
Can this be a `if (!r) continue` rather than indenting the whole block an extra time?